docs/types/literal-suffixes.rst

```diff
@@ -0,0 +1,243 @@
+there are types for which no built-in conversions are available (most notably the
+:ref:`user-defined value types<user-defined-value-types>`).
+
+To fill this gap, it is possible to define custom conversions as *literal suffixes*.
+
+.. code-block:: solidity
+    :force:
+
+    // SPDX-License-Identifier: GPL-3.0
+    pragma solidity ^0.8.20;
+
+    type Coin is uint;
+    using {add as +, eq as ==} for Coin global;
+
+    function add(Coin a, Coin b) pure returns (Coin) {
+        return Coin.wrap(Coin.unwrap(a) + Coin.unwrap(b));
+    }
+
+    function eq(Coin a, Coin b) pure returns (bool) {
+        return Coin.unwrap(a) == Coin.unwrap(b);
+    }
+
+    function c(uint mantissa, uint exponent) pure suffix returns (Coin) {
+        return Coin.wrap(mantissa * 10**6 / 10**exponent);
+    }
+
+    function uc(uint microValue) pure suffix returns (Coin) {
+        return Coin.wrap(microValue);
+    }
+
+    contract CoinSeller {
+        mapping(address => Coin) balances;
+
+        function buy() public payable {
+            require(msg.value == 1 ether);
+            assert(1.5 c == 1_500_000 uc);
+            balances[msg.sender] = balances[msg.sender] + 1.5 c;
+        }
+    }
+
+.. index:: ! literal suffix;definition, function;free
+
+Defining Suffix Functions
+-------------------------
+
+Literal suffixes can be defined by applying the built-in ``suffix`` modifier to a :ref:`free function<functions>`.
+
+Only pure functions can be used as suffixes.
+This means that suffixes cannot read or modify blockchain state.
+As with all pure functions, however, they can perform pure external calls.
+
+.. index:: literal;address
+
+Suffix Function Parameters
+^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+A suffix function must accept and return exactly one value.
+As a special case, suffixes on :ref:`rational literals<rational_literals>` can optionally accept two arguments,
+produced by the :ref:`fractional decomposition<fractional_decomposition>` of such a literal.
+
+Suffixes can only have parameters of types for which an implicit conversion from a literal exists.
+For single-parameter suffixes, this includes the following types:
+
++-------------------------------------------------------------+----------------------------------------------------------------+
+| Parameter type                                              | Accepted literals                                              |
++=============================================================+================================================================+
+| ``bool``                                                    | - :ref:`Boolean<booleans>` literals                            |
++-------------------------------------------------------------+----------------------------------------------------------------+
+| ``uint8``, ..., ``uint256``, ``int8``, ..., ``int256``      | - :ref:`Rational literals<rational_literals>` (including zero) |
```

Comment:
why is it so special here to write including zero? I would have expected that zero is included.

Reply:
Because zero is a special case. For example `1` is not a valid value for `bytes4`, but `0` is. This is the only rational number literal that can be assigned to any fixed-bytes type.

You might expect to be able to use fixed-bytes suffixes on `0x00`, but not necessarily on `0`.

Reply:
Also, it's specifically zero as a rational number. An address literal with zero value does not work that way. I don't want to overload the table with details, but at least the fact that zero is special warrants a mention.